Add Jonke synapse

weidel-p opened the pull request:
@jeanningrosloic and I implemented the synapse model as described in [Feedback Inhibition Shapes Emergent Computational
Properties of Cortical Microcircuit Motifs](https://www.jneurosci.org/content/37/35/8511). 

We think that this synapse model could be useful for the community, please consider adding it to NEST.

weidel-p commented on pynest/nest/tests/test_NodeCollection.py:
```diff
@@ -518,7 +518,7 @@ def test_GetConnections(self):
         self.assertEqual(get_conn_some.get('target'), compare_target)
 
-        compare_list = [3, 1, 0, 15, 6]
+        compare_list = [3, 1, 0, 18, 6]
```
Here and in `test_synapsecollection.py` I had to change the synapse_id of the static connection from 15 to 18. Apparently, by adding a synapse model the IDs of the other synapses change which leads to failing tests.
Is there a way to automatically get the correct synapse id for the static connection instead of hard-coding it?

clinssen replied:
@hakonsbm: could you comment on whether a hard-coded ID is OK here, or whether we need to dynamicaly get this from the kernel?

hakonsbm replied:
@clinssen If all synapse IDs change when adding a synapse, we should get the synapse ID for reference values dynamically, so don't have to change the reference values every time a new synapse is added. The synapse ID of a given synapse model can be found in the synapsedict, so we could do something like this for example:
```python
expected_syn_model = 'static_synapse'
expected_syn_id = nest.ll_api.sli_func('synapsedict')[expected_syn_model]
compare_list = [3, 1, 0, expected_syn_id, 6]
```
And similarly in other places we use synapse IDs as reference values.

weidel-p replied:
I did what you suggested. It seems to work.

clinssen commented on pynest/nest/tests/test_jonke_synapse.py:
```diff
@@ -0,0 +1,269 @@
+        self.assertAlmostEqual(
+            weight_reproduced_independently,
+            weight_by_nest,
+            msg=f"{self.synapse_parameters['synapse_model']} test:\n" +
+                f"Resulting synaptic weight {weight_by_nest} " +
+                f"differs from expected {weight_reproduced_independently}")
```
Good catch, but now superseded by #1704.

weidel-p replied:
Do I understand correctly, that the f-strings are not a problem?

hakonsbm replied:
Yes, they can be left as they are.

weidel-p commented on pynest/nest/tests/test_jonke_synapse.py:
```diff
@@ -0,0 +1,269 @@
+        t_previous_post = -1
+        syn_trace_pre = 0.0
+        syn_trace_post = 0.0
+        w = _initial_weight
```
done

weidel-p commented on pynest/nest/tests/test_jonke_synapse.py:
```diff
@@ -0,0 +1,269 @@
+import unittest
+import nest
+import numpy as np
+import matplotlib.pyplot as plt
```
done

weidel-p commented:
I think I addressed all the comments so far. Now the CI is failing again and I don't understand why.

The error Travis is throwing:
Failure: OSError (No such file /home/travis/build/nest/nest-simulator/build/nest/tests) ... ERROR

This seems unrelated to this PR, maybe this is an issue with Travis itself?

hakonsbm commented:
@weidel-p It's a problem with the installation on Travis and currently affects all PRs. I have created #1743.

clinssen commented:
@hakonsbm: the Travis issue has been fixed, could you take another look? Cheers!

hakonsbm commented:
@weidel-p There are still some suggestions not yet addressed. Remember to click "**Load more...**" to see all my suggestions.

stinebuu commented:
@weidel-p have you had time to look at the rest of @hakonsbm suggestions? Thanks!

weidel-p commented on pynest/nest/tests/test_jonke_synapse.py:
```diff
@@ -0,0 +1,269 @@
+            "tau_minus": 33.7
+        }
+
+    def test_weight_drift(self):
```
done

weidel-p commented:
Yes, I did now. Please take a look.

hakonsbm reviewed (changes requested):
I have a few more suggestions. And `spike_detector` is now renamed to `spike_recorder`, which is why Travis complains.

weidel-p commented:
@hakonsbm I followed all your requested changes. Please take a look again.

hakonsbm commented on pynest/nest/tests/test_jonke_synapse.py:
```diff
@@ -0,0 +1,263 @@
+        "tau_minus": 33.7
+    }
+
+    def setUp(self):
+        pass
+
```
As it's empty, we can remove the `setUp()`
```suggestion
```

weidel-p commented:
@hakonsbm I removed the setUp function.

hakonsbm reviewed (approved):
Great, thanks!
